_posts/2021-02-01-aws-config-compliance-automation.md

```diff
@@ -0,0 +1,95 @@
+
+If ServiceNow is accurately reflecting configuration item (CI) data in the CMDB, there are some useful direct integrations possible using a combination of AWS Config, AWS Simple Notification Service (SNS), and ServiceNow. However, many organizations do not have a 1:1 reflection of AWS resources as CI's in their ServiceNow CMDB, so some customization is required. Unfortunately when this is necessary, the <a href="https://docs.servicenow.com/bundle/paris-it-operations-management/page/product/cloud-management-v2-setup/task/aws-config-service-cloud-mgt.html">out-of-the-box capabilities utilizing the tools</a> is not possible.
+
+## Alternative solution
+
+In these cases, some customization is necessary to drive visibility of the non-compliant items from AWS Config into the normal resolution workflow in ServiceNow. A solution that is possible, is to pull data from AWS Config using `aws-sdk` API's, and then feeding that information into ServiceNow using ServiceNow API's.
+
+Many organizations deploy resources into several AWS accounts, making management of these items a little trickier. Luckily, AWS Config provides a capability to track compliance data at an <a href="https://docs.aws.amazon.com/config/latest/developerguide/aggregate-data.html">aggregate-level, across many accounts</a>. Once aggregated, the data is made available in a few different ways, but the most flexible seems to be utilizing <a href="https://docs.aws.amazon.com/config/latest/developerguide/querying-AWS-resources.html">AWS Config Advanced Queries</a> to pull back compliance information across many accounts while utilizing the same method to get the data for each. 
+
+A sample query used from the aggregator to get non-compliant information from a single account could look like the following:
+
+```sql
+SELECT
+  resourceId,
+  configuration.targetResourceType,
+  configuration.complianceType,
+  configuration.configRuleList
+WHERE
+  accountId = '<account-id>'
+  AND configuration.complianceType = 'NON_COMPLIANT'
+```
+
+which pulls back the offending resource, its type, and information about which AWS Config rules the resource was shown to be non-compliant for.
+
+Once this information is retrieved, it can be compared against previous known state to determine if the compliance finding is new (and needing ServiceNow tracking) or has been previously reported (and subsequently tracked in ServiceNow). It can also be determined if a finding has been previously reported, but the AWS resource is now shown to be back in compliance (say for instance, if someone in IT has manually intervened to reconfigure the resource such that it now complies with configuration rules). If the resource is back in compliance, the correlated ServiceNow item used to track the offense can be successfully closed.
```

Comment:
On the middle sentence, maybe, "It can also be used to determine if a resource, which was previously reported as non compliant, is now compliant (say for instance..."